src/lib/__tests__/SpeculationEngine.test.ts

```diff
@@ -0,0 +1,284 @@
+    pullRequest,
+  }),
+  requestId: '0',
+  state: 'queued',
+});
+
+const landRequestC = new LandRequestStatus({
+  date: new Date(120),
+  id: '0',
+  isLatest: true,
+  request: new LandRequest({
+    created: new Date(120),
+    forCommit: 'abc',
+    id: '0',
+    triggererAaid: '123',
+    pullRequestId: 0,
+    pullRequest,
+  }),
+  requestId: '0',
+  state: 'running',
+});
+const landRequestD = new LandRequestStatus({
+  date: new Date(120),
+  id: '1',
+  isLatest: true,
+  request: new LandRequest({
+    created: new Date(120),
+    forCommit: 'abc',
+    id: '1',
+    triggererAaid: '123',
+    pullRequestId: 1,
+    pullRequest,
+  }),
+  requestId: '0',
+  state: 'running',
+});
+
+describe('SpeculationEngine', () => {
+  beforeEach(() => {
+    mockPullRequest = {
+      pullRequestId: 1,
+      authorAaid: '123',
+      title: 'Foo',
+      sourceBranch: 'test',
+      targetBranch: 'master',
+      commit: 'abc',
+    } as BB.PullRequest;
+
+    mockQueued = [landRequestA, landRequestB];
+    mockRunning = [landRequestC, landRequestD];
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+  test('getAvailableSlots > should return available free slots', async () => {
+    jest.spyOn(StateService, 'getMaxConcurrentBuilds').mockResolvedValue(3);
+    expect(await SpeculationEngine.getAvailableSlots(mockRunning)).toBe(1);
+    expect(await SpeculationEngine.getAvailableSlots([])).toBe(3);
+  });
+
+  test('positionInQueue > should return position in queue', () => {
+    expect(SpeculationEngine.getPositionInQueue(mockQueued, mockQueued[1])).toBe(1);
+    expect(SpeculationEngine.getPositionInQueue(mockQueued, mockQueued[0])).toBe(0);
+  });
+
+  test('getLowestImpactedRequestStatus > should return the lowest impact land request', () => {
+    const impact = SpeculationEngine.getLowestImpactedRequestStatus(mockQueued, 0, 2);
+    expect(impact).toEqual(landRequestB);
+  });
+
+  describe('When a reorder request is successful', () => {
+    test('should reorder when number of free slots are 2, current request is 1st in the queue and current request`s impact is greater than next', async () => {
+      jest.spyOn(StateService, 'getAdminSettings').mockResolvedValue({
+        speculationEngineEnabled: true,
+      } as any);
+      jest.spyOn(StateService, 'getMaxConcurrentBuilds').mockResolvedValueOnce(3);
+      // 2 free slots, getMaxConcurrentBuilds is 3 and running is 1
+      // queued length is 2
+      // position in queue is 0 ie 1st in the queue
+      jest.spyOn(SpeculationEngine, 'getLowestImpactedRequestStatus').mockReturnValue(landRequestB);
+
+      const reorder = await SpeculationEngine.reorderRequest(
+        mockRunning.slice(0, 1),
+        mockQueued.slice(0, 2),
+        landRequestA,
+      );
+
+      expect(SpeculationEngine.getLowestImpactedRequestStatus).toHaveBeenCalledWith(
+        mockQueued.slice(0, 2),
+        0,
+        2,
+      );
+      expect(reorder).toBe(true);
+    });
+  });
```

Comment:
Do we want to add a successful case for 3 slots as well?

Reply:
Yeah, can add another test case.